### PR TITLE
Set the entitlements to cancelled when program is archived

### PR DESCRIPTION
**Pull request by gonzalesedwin1123**

### CODE REVIEW

File: g2p_programs/models/managers/entitlement_manager.py [hunk 1]
```diff
@@ -60,6 +63,70 @@ def approve_entitlements(self, entitlements):
...
+        """
+        Base Entitlement Manager :meth:`_cancel_entitlements`
+        Synchronous cancellation of entitlements in a cycle
+        Override in entitlement manager
+
+        :param cycle: A recordset of cycle
+        :param offset: An integer value to be used in :meth:`cycle.get_entitlements` for setting the query offset
+        :param limit: An integer value to be used in :meth:`cycle.get_entitlements` for setting the query limit
+        :return:
+        """
```
Comment (jeremi) [inline g2p_programs/models/managers/entitlement_manager.py:220]: Is `pass` missing ?

File: g2p_programs/models/programs.py [hunk 2]
```diff
@@ -112,6 +112,31 @@ def _default_journal_id(self):
...
+        :return: toggle_active function of parent class
+        """
+        for rec in self:
+            # Cancel cycles and entitlements only if the program is active (for archiving)
+            if rec.active:
+                _logger.info("Archive Program: cancel cycles and entitlements.")
+                if rec.cycle_ids:
+                    entitlement_manager = rec.get_manager(self.MANAGER_ENTITLEMENT)
+                    # Get only `draft` and `to_approve` cycles
+                    cycles = rec.cycle_ids.filtered(
```
Comment (jeremi) [inline g2p_programs/models/programs.py:134]: Shouldn't we deactivate all entitlements even though the cycle is active ?

File: g2p_programs/models/managers/entitlement_manager.py [hunk 3]
```diff
@@ -224,6 +291,43 @@ def prepare_entitlements(self, cycle, beneficiaries):
...
+            ["draft", "pending_validation"],
+            entitlement_model="g2p.entitlement",
+            count=True,
+        )
+        if entitlements_count < self.MIN_ROW_JOB_QUEUE:
+            self._cancel_entitlements(cycle)
+        else:
+            self._cancel_entitlements_async(cycle, entitlements_count)
+
+    def _cancel_entitlements(self, cycle, offset=0, limit=None):
```
Comment (jeremi) [inline g2p_programs/models/managers/entitlement_manager.py:529]: I think it should also disable entitlements that have been validated but not redeemed.

**Review by jeremi — CHANGES_REQUESTED** [review]

File: g2p_programs/models/programs.py (hunk 2 above)
Comment (gonzalesedwin1123) [reply g2p_programs/models/programs.py:134]: Do you mean deactivate entitlements with approved cycles?

Comment (gonzalesedwin1123) [reply g2p_programs/models/programs.py:134]: Your comment below answered by question. I will work on this now.

File: g2p_programs/models/managers/entitlement_manager.py (hunk 1 above)
Comment (gonzalesedwin1123) [reply g2p_programs/models/managers/entitlement_manager.py:220]: Fixed

File: g2p_programs/models/programs.py (hunk 2 above)
Comment (gonzalesedwin1123) [reply g2p_programs/models/programs.py:134]: Fixed

File: g2p_programs/models/managers/entitlement_manager.py (hunk 3 above)
Comment (gonzalesedwin1123) [reply g2p_programs/models/managers/entitlement_manager.py:529]: Fixed

File: g2p_programs/models/managers/entitlement_manager.py (hunk 1 above)
Comment (gonzalesedwin1123) [reply g2p_programs/models/managers/entitlement_manager.py:220]: pass is removed by pre-commit in function without codes. I changed it to `raise NotImplementedError()`

**Review by jeremi — APPROVED** [review]

**Review by shibu-narayanan — APPROVED** [review]

**Pull request merged**